transformer_engine/common/comm_gemm_overlap/userbuffers-host.cpp

```diff
@@ -0,0 +1,1198 @@
+
+  if (ndev == numlocal) {  // all visible devices
+    if (cur_dev != mylocal)
+      printf("%d: device used %d[%d] ,resetting device to %d\n", rank, cur_dev, ndev, mylocal);
+    CUDACHECK(cudaSetDevice(mylocal));
+  }
+  (*comm)->mydev = cur_dev;
+  // FIXME need to check that numlocal is multiple of pipegpus x tensorgpus
+  // ar1 is data
+  int divgpus = pipegpus * tensorgpus;
+  int datagpus = numlocal / divgpus;
+  (*comm)->ar_nvsize = datagpus;
+  (*comm)->ar_firstgpu = mylocal - ((mylocal / tensorgpus) % datagpus) * tensorgpus;
+  (*comm)->ar_nvrank = (mylocal - (*comm)->ar_firstgpu) / tensorgpus;
+  // ar2 is tensor
+  (*comm)->ar2_nvsize = tensorgpus;
+  (*comm)->ar2_firstgpu = mylocal - mylocal % tensorgpus;
+  (*comm)->ar2_nvrank = mylocal - (*comm)->ar2_firstgpu;
+  // ar2 has step equal to ar_nvsize
+  int allnodes = nranks / numlocal;
+  int mynode = myrank / numlocal;
+  int datanodes = allnodes / pipenodes / tensornodes;
+  int pipenodegroup_id = myrank / numlocal / (datanodes * tensornodes);
+
+  (*comm)->pipe_id = pipegpus * pipenodegroup_id + mylocal / (datagpus * tensorgpus);
+
+  CUDACHECK(cudaFree(0));
+  int datanodegroup_id =
+      myrank / numlocal / datanodes;  // data reduction group node belongs, equals 0 for all if both
+                                      // pipenodes=1 and tensornodes=1
+  // mpi communicator only needed for SHARP which is always allreduce1/data-parallel
+  MPI_Comm_split(MPI_COMM_WORLD, mylocal + numlocal * datanodegroup_id, rank, &(*comm)->comm_inter);
+  // different rails from same group are in different subcommunicators
+
+  MPI_Comm_size((*comm)->comm_inter, &num_nodes);
+  MPI_Comm_rank((*comm)->comm_inter, &my_node);
+  (*comm)->first_node = mynode - my_node;
+  (*comm)->num_nodes = num_nodes;
+  (*comm)->my_node = my_node;
+
+  (*comm)->num2_nodes = tensornodes;
+  (*comm)->my2_node = (mynode / datanodes) % tensornodes;
+  (*comm)->first2_node = mynode - (*comm)->my2_node * datanodes;
+
+  char *ib_dev_list;
+  int ZIONROCE = getenv("ZIONROCE") ? atoi(getenv("ZIONROCE")) : 0;
+  int ROCE = getenv("ROCE") ? atoi(getenv("ROCE")) : 0;
+  if (ZIONROCE) ROCE = 1;
+  int PREOS = getenv("PREOS") ? atoi(getenv("PREOS")) : 0;
+
+  switch (mylocal) {
+		case 0:ib_dev_list = "mlx5_0:1"; break;  // NOLINT(*)
+		case 1:ib_dev_list = (char*)(PREOS?"mlx5_3:1":"mlx5_1:1"); break;  // NOLINT(*)
+		case 2:ib_dev_list = (char*)(ZIONROCE?"mlx5_4:1":PREOS?"mlx5_4:1":"mlx5_2:1"); break;  // NOLINT(*)
+		case 3:ib_dev_list = (char*)(PREOS?"mlx5_5:1":"mlx5_3:1"); break;  // NOLINT(*)
+		case 4:ib_dev_list = (char*)(PREOS?"mlx5_6:1":"mlx5_6:1"); break;  // NOLINT(*)
+		case 5:ib_dev_list = (char*)(PREOS?"mlx5_9:1":"mlx5_7:1"); break;  // NOLINT(*)
+		case 6:ib_dev_list = (char*)(ZIONROCE?"mlx5_10:1":PREOS?"mlx5_10:1":"mlx5_8:1"); break;  // NOLINT(*)
+		case 7:ib_dev_list = (char*)(PREOS?"mlx5_11:1":"mlx5_9:1"); break;  // NOLINT(*)
+                default: break;
+  }
```

Comment:
1. Use getCudaDeviceProperties to query major and minor versions.
2. Find compute capability and then switch on those.

Reply:
Done. Please check.